cmd/podman/images/scp.go

```diff
@@ -0,0 +1,344 @@
+		ind = 1
+	}
+	cfg, err := connection.ValidateAndConfigure(scpOpts.URI[ind], scpOpts.Iden[ind])
+	if err != nil {
+		return nil, "", err
+	}
+	dialAdd, err := ssh.Dial("tcp", scpOpts.URI[ind].Host, cfg) // dial the client
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "failed to connect")
+	}
+	file, err := makeRemoteFile(dialAdd)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return dialAdd, file, nil
+}
+
+// validateImageName makes sure that the image given is valid and no injections are occurring
+// we simply use this for error checking, bot setting the image
+func validateImageName(input string) error {
+	// ParseNormalizedNamed transforms a shortname image into its
+	// full name reference so busybox => docker.io/library/busybox
+	// we want to keep our shortnames, so only return an error if
+	// we cannot parse what th euser has given us
+	_, err := reference.ParseNormalizedNamed(input)
+	return err
+}
+
+// parseArgs returns the valid connection data based off of the information provided by the user
+// args is an array of the command arguments and cfg is tooling configuration used to get service destinations
+// returned is serv and an error if applicable. serv is a map of service destinations with the connection name as the index
+// this connection name is intended to be used as EngineConfig.ServiceDestinations
+// this function modifies the global scpOpt entities: FromRemote, ToRemote, Connections, and SourceImageName
+func parseArgs(args []string, cfg *config.Config) (map[string]config.Destination, error) {
+	iden := []string{}
+	url := []string{}
+	serv := map[string]config.Destination{}
+	cliConnections := []string{}
+	switch len(args) {
+	case 1:
+		if strings.Contains(args[0], "::") {
+			scpOpts.FromRemote = true
+			cliConnections = append(cliConnections, args[0])
+		} else {
+			err := validateImageName(args[0])
+			if err != nil {
+				return nil, err
+			}
+			scpOpts.SourceImageName = args[0]
+		}
+	case 2:
+		if strings.Contains(args[0], "::") {
+			if !(strings.Contains(args[1], "::")) && len(strings.Split(args[0], "::")[1]) <= 1 { // if an image is specified, this mean we are loading from our client
+				cliConnections = append(cliConnections, args[0])
+				scpOpts.ToRemote = true
+			} else if strings.Contains(args[1], "::") { // both remote clients
+				scpOpts.FromRemote = true
+				scpOpts.ToRemote = true
+				if len(strings.Split(args[0], "::")[1]) <= 1 { // is save->load w/ one image name or save-load w/ two image names
+					cliConnections = append(cliConnections, args[0])
+					cliConnections = append(cliConnections, args[1])
+				} else if len(strings.Split(args[0], "::")[1]) > 1 && len(strings.Split(args[1], "::")[0]) > 1 {
+					return nil, errors.Wrapf(define.ErrInvalidArg, "cannot specify an image rename")
+				} else { // else its a load save (order of args)
+					cliConnections = append(cliConnections, args[1])
+					cliConnections = append(cliConnections, args[0])
```

Comment:
I don’t see how the switching of order of arguments works — later the `createConnection` function assumes that the index in `.Connections` depends only on direction.

Reply:
this is so `fed::alpine fed::` and `fed:: fed::alpine` both mean the same thing. cliConnections/scpOpts.Connections always has the first entry be the `from` location and the second be the `to` location

Reply:
In that case of copying from/to the same host, sure, that works, but for `src::alpine dest::` and `src:: dest::alpine` this breaks, AFAICS, switching `src` and `dest`, exactly because “the first entry is the from location”.

Reply:
I don't see how the example using different hosts makes a difference. the ordering paradigm of first entry in array = from location and second entry in array = to location still holds

Reply:
I might have misunderstood the design, see elsewhere. Anyway, assuming my previous interpretation:

- `image scp a:: b::alpine`:
  -  My expectation of what this means: load from connection `a`, save to connection `b`.
  - I think the code does: The `// is save->load w/ one image name or save-load w/ two image names` code path, sets `cliConnections` to `["a::alpine", "b::"]`, i.e. `scpOpts.URI` is set to `[aConnection, bConnection]`, i.e. from=`a`, to=`b`
  - i.e. works as I thought it is designed to
- `image scp a::alpine b::`:
  -  My expectation of what this means: load from connection `a`, save to connection `b`.
  - I think the code does: The `// else its a load save (order of args)` code path, sets `cliConnections` to `["b::" ,"a::alpine"]`, i.e. `scpOpts.URI` is set to `[bConnection, aConnection]`, i.e. from=`b`, to=`a`
  - i.e. works contrary to my expectations